### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): closed immersions with affine target

### PR DESCRIPTION
**Pull request by chrisflav**
Shows that closed immersions are affine morphisms inducing affine locally surjective maps on sections.

---

- [x] depends on: #15033 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/AffineScheme.lean [hunk 1]
```diff
@@ -712,6 +735,104 @@ lemma Scheme.eq_zeroLocus_of_isClosed_of_isAffine (X : Scheme.{u}) [IsAffine X]
 
 end ZeroLocus
 
+section Factorization
+
+variable {X : Scheme.{u}} {A : CommRingCat}
+
+/-- If `X ⟶ Spec A` is a morphism of schemes, then `Spec` of this ring is the scheme-theoretic
```
Comment (MichaelStollBayreuth) [inline Mathlib/AlgebraicGeometry/AffineScheme.lean:742]:
```suggestion
/-- If `f : X ⟶ Spec A` is a morphism of schemes, then `Spec A` is the scheme-theoretic
```
Is this the intended meaning?

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/AffineScheme.lean:742]: Ah, "this ring" is the ring `specTargetImage f`, i.e. this refers to the ring defined in this definition.

File: Mathlib/AlgebraicGeometry/AffineScheme.lean [hunk 2]
```diff
@@ -712,6 +735,104 @@ lemma Scheme.eq_zeroLocus_of_isClosed_of_isAffine (X : Scheme.{u}) [IsAffine X]
...
+def specTargetImage (f : X ⟶ Spec A) : CommRingCat :=
+  CommRingCat.of (A ⧸ (RingHom.ker <| (((ΓSpec.adjunction).homEquiv X (op A)).symm f).unop))
+
+/-- If `f : X ⟶ Spec A` is a morphism of schemes, then `f` factors via
+the inclusion of `Spec (specTargetImage f)` into `X`. -/
+def specTargetImageFactorization (f : X ⟶ Spec A) : X ⟶ Spec (specTargetImage f) :=
+  (ΓSpec.adjunction).homEquiv X (op <| specTargetImage f) (Opposite.op (RingHom.kerLift _))
+
+/-- If `f : X ⟶ Spec A` is a morphism of schemes, the induced morphism on spectra of this
+ring homomorphism is the inclusion of the scheme-theoretic image of `f` into `Spec A`. -/
```
Comment (MichaelStollBayreuth) [inline Mathlib/AlgebraicGeometry/AffineScheme.lean:755]: Maybe also replace "this ring homomorphism" by "`specTargetImageRingHom f`".

File: Mathlib/AlgebraicGeometry/AffineScheme.lean [hunk 3]
```diff
@@ -578,6 +578,15 @@ theorem isLocalization_stalk (x : U) :
   subst this
   exact hU.isLocalization_stalk' y hx
 
+lemma stalkMap_injective_of (f : X ⟶ Y) {U : Opens Y} (hU : IsAffineOpen U) (x : X)
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/AffineScheme.lean:581]: I don't think the trailing `_of` is necessary? Same for `stalkMap_injective_of_isAffine_of`.

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 4]
```diff
@@ -465,6 +465,22 @@ lemma ΓSpec_adjunction_homEquiv_eq {X : Scheme.{u}} {B : CommRingCat} (φ : B 
   simp_rw [← ΓSpecIso_inv_ΓSpec_adjunction_homEquiv φ]
   simp
 
+lemma toSpecΓ_ΓSpec_adjunction_homEquiv {X : Scheme.{u}} {B : CommRingCat} (φ : B ⟶ Γ(X, ⊤)) :
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:468]: Is this just `ΓSpecIso_inv_ΓSpec_adjunction_homEquiv`?
What's the right spelling here? I'm thinking we should just remove `toSpecΓ`.

File: Mathlib/AlgebraicGeometry/Morphisms/QuasiSeparated.lean [hunk 5]
```diff
@@ -347,6 +347,32 @@ theorem is_localization_basicOpen_of_qcqs {X : Scheme} {U : X.Opens} (hU : IsCom
...
     simpa [mul_comm z] using e
 
+lemma exists_of_res_eq_of_qcqs {X : Scheme.{u}} {U : TopologicalSpace.Opens X}
+    (hU : IsCompact U.carrier) (hU' : IsQuasiSeparated U.carrier)
+    {f g s : Γ(X, U)} (hfg : f |_ X.basicOpen s = g |_ X.basicOpen s) :
+    ∃ n, s ^ n * f = s ^ n * g := by
+  obtain ⟨n, hc⟩ := (is_localization_basicOpen_of_qcqs hU hU' s).exists_of_eq s hfg
+  use n
+
+lemma exists_of_Γres_eq_of_qcqs {X : Scheme.{u}} [CompactSpace X] [QuasiSeparatedSpace X]
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/QuasiSeparated.lean:357]: I'm not entirely sure about the name. Maybe just `_of_top`.

File: Mathlib/AlgebraicGeometry/Scheme.lean [hunk 6]
```diff
@@ -92,6 +92,8 @@ namespace Hom
 
 variable {X Y : Scheme.{u}} (f : Hom X Y) {U U' : Y.Opens} {V V' : X.Opens}
 
+lemma map_le (hUU' : U ≤ U') : f⁻¹ᵁ U ≤ f⁻¹ᵁ U' := fun _ ha ↦ hUU' ha
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Scheme.lean:95]:
```suggestion
lemma preimage_le (hUU' : U ≤ U') : f⁻¹ᵁ U ≤ f⁻¹ᵁ U' := fun _ ha ↦ hUU' ha
```

File: Mathlib/AlgebraicGeometry/Scheme.lean [hunk 7]
```diff
@@ -581,6 +583,14 @@ theorem Scheme.Spec_map_presheaf_map_eqToHom {X : Scheme} {U V : X.Opens} (h : U
   refine (Scheme.congr_app this _).trans ?_
   simp [eqToHom_map]
 
+lemma germ_eq_zero_of {X : Scheme.{u}} {U : Opens X} (x : U) {f s : Γ(X, U)}
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Scheme.lean:586]:
```suggestion
lemma germ_eq_zero_of_pow_mul_eq_zero {X : Scheme.{u}} {U : Opens X} (x : U) {f s : Γ(X, U)}
```
Maybe?

File: Mathlib/Geometry/RingedSpace/Basic.lean [hunk 8]
```diff
@@ -41,12 +41,27 @@ namespace RingedSpace
...
+
 variable (X : RingedSpace)
 
 -- Porting note (#10670): this was not necessary in mathlib3
 instance : CoeSort RingedSpace Type* where
   coe X := X.carrier
 
+/-- If the germ of a section `f` is zero in the stalk at `x`, then `f` is zero on some neighbourhood
+around `x`. -/
+lemma eq_zero_res_of_eq_zero_germ (U : Opens X) (f : X.presheaf.obj (op U)) (x : U)
```
Comment (erdOne) [inline Mathlib/Geometry/RingedSpace/Basic.lean:57]:
```suggestion
lemma res_eq_zero_of_germ_eq_zero (U : Opens X) (f : X.presheaf.obj (op U)) (x : U)
```
(or maybe `exists_`?)

File: Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean [hunk 9]
```diff
@@ -68,6 +68,12 @@ lemma evaluation_ne_zero_iff_mem_basicOpen (x : U) (f : X.presheaf.obj (op U)) :
     X.evaluation x f ≠ 0 ↔ x.val ∈ X.toRingedSpace.basicOpen f := by
   simp
 
+lemma basicOpen_eq_bot_iff_forall_evaluation_eq_zero (f : X.presheaf.obj (op U)) :
+    X.toRingedSpace.basicOpen f = ⊥ ↔ ∀ (x : U), X.evaluation x f = 0 := by
+  simp only [evaluation_eq_zero_iff_not_mem_basicOpen, Subtype.forall]
+  exact ⟨fun h ↦ h ▸ fun a _ hc ↦ hc,
+    fun h ↦ eq_bot_iff.mpr <| fun a ha ↦ h a (X.toRingedSpace.basicOpen_le f ha) ha⟩
+
```
Comment (erdOne) [inline Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:76]: Can you specialise this to schemes as well?

File: Mathlib/RingTheory/Localization/Basic.lean [hunk 10]
```diff
@@ -546,6 +546,14 @@ theorem lift_injective_iff :
     Injective (lift hg : S → P) ↔ ∀ x y, algebraMap R S x = algebraMap R S y ↔ g x = g y :=
   (toLocalizationMap M S).lift_injective_iff hg
 
+variable (M) in
+include M in
+lemma injective_iff {T} [CommRing T] (f : S →+* T) :
```
Comment (erdOne) [inline Mathlib/RingTheory/Localization/Basic.lean:551]: I think `IsLocalization.injective_iff` is a bit too generic for this. Maybe `injective_iff_map_algebraMap_eq` or something else that mentions `algebraMap`. Same for `injective_of`.

File: Mathlib/AlgebraicGeometry/Morphisms/QuasiCompact.lean [hunk 11]
```diff
@@ -280,4 +287,11 @@ lemma Scheme.zeroLocus_eq_top_iff_subset_nilradical_of_isCompact {X : Scheme.{u}
   simp [Scheme.zeroLocus_def, ← Scheme.isNilpotent_iff_basicOpen_eq_bot_of_isCompact hU,
     ← mem_nilradical, Set.subset_def]
 
+/-- The zero locus of a set of sections over a compact open of a scheme is `X` if and only if
+`s` is contained in the nilradical of `Γ(X, U)`. -/
+lemma Scheme.zeroLocus_eq_top_iff_subset_nilradical_of_compactSpace {X : Scheme.{u}}
+    [CompactSpace X] (s : Set Γ(X, ⊤)) :
+    X.zeroLocus s = ⊤ ↔ s ⊆ (nilradical Γ(X, ⊤)).carrier :=
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/QuasiCompact.lean:294]:
```suggestion
    X.zeroLocus s = ⊤ ↔ s ⊆ nilradical Γ(X, ⊤) :=
```
Does this not work?

File: Mathlib/AlgebraicGeometry/Morphisms/QuasiCompact.lean [hunk 12]
```diff
@@ -272,6 +272,13 @@ lemma Scheme.isNilpotent_iff_basicOpen_eq_bot_of_isCompact {X : Scheme.{u}}
   rw [mul_one] at hn
   use n
 
+/-- A global section of a quasi-compact scheme is nilpotent if and only if its associated
+basic open is empty. -/
+lemma Scheme.isNilpotent_iff_basicOpen_eq_bot_of_compactSpace {X : Scheme.{u}}
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/QuasiCompact.lean:277]: I feel that we can drop the `_of_compactSpace`; same below.

**Review by erdOne — COMMENTED** [review]
Some bikeshedding on names. Not yet looked into the main bits.

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean (hunk 4 above)
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:468]: Good catch, this was accidentally re-introduced by merging.

**Comment by erdOne** [comment]
Can you merge master and incorporate the new changes in the refactor? Thanks!

**Review by erdOne — APPROVED** [review]
Otherwise LGTM

**Comment by chrisflav** [comment]
> Can you merge master and incorporate the new changes in the refactor? Thanks!

Done.

File: Mathlib/RingTheory/Localization/Defs.lean [hunk 13]
```diff
@@ -528,6 +528,14 @@ theorem lift_injective_iff :
     Injective (lift hg : S → P) ↔ ∀ x y, algebraMap R S x = algebraMap R S y ↔ g x = g y :=
   (toLocalizationMap M S).lift_injective_iff hg
 
+variable (M) in
+include M in
```
Comment (erdOne) [inline Mathlib/RingTheory/Localization/Defs.lean:532]: Is this how it works?

**Comment by erdOne** [comment]
Thanks a lot!
maintainer merge

**Comment by riccardobrasca** [comment]
Modulo Andrew's comment LGTM, thanks!

bors d+

File: Mathlib/RingTheory/Localization/Defs.lean (hunk 13 above)
Comment (erdOne) [reply Mathlib/RingTheory/Localization/Defs.lean:532]: Sure, let's merge it then.

Comment (chrisflav) [reply Mathlib/RingTheory/Localization/Defs.lean:532]: It seems there is no shorter way, see https://leanprover.zulipchat.com/#narrow/channel/113488-general/topic/include.20in.20with.20variable.20in.

**Comment by chrisflav** [comment]
Thanks for the reviews!

bors r+

**Pull request closed without merging**